### DISCUSSION TITLE
Executing commands locally from the machine bot is hosted on

### DISCUSSION THREAD

**Discussion by weeeeelp** [jagrosh/MusicBot#1173] [Ideas]
Hey, first of all thanks for this piece of software, we're using the bot for quite some time now and we're very happy with it.

I was wondering about the possibility of implementing a way to execute bot commands not just from a Discord text channel, but also locally from the machine the bot is running on. 

Let me quickly describe the use case:
- we have several hundred of mp3 files that can be played by users 
- they execute a command like `-play dir/file.mp3`
- sadly the user still needs to find the right one on the list, copy it and paste into the text chat to trigger JMusicBot to play the sound file

I would like to make that less painful by writing a web UI for people to click the desired file on the list and have it automatically played (I would do that myself, obviously). However, this would require a way to make the bot accept commands from "outside" - not just the Discord text channels, but for example a locally-bound port that would accept some payload with a command to execute. A curl mock for that would be something like:

`curl -H 'Content-Type: application/json' -d '{"discord_userid":"317791173500xxxxxx", "command": "play dir/file.mp3"}' 127.0.0.1:1234 `

executed locally from the machine JMusicBot is running on. 
This should cause the bot to play the sound in the voice channel the discord_userid is connected to.

Obviously, this doesn't have to be limited to just -play, could be any number of commands. This could bring benefits to people looking to do some external integration with the bot, obviously due to security considerations should be disabled by default and maybe limited to localhost only. Not hellbent on bound port/json either - there are many mechanisms for that, just wondering what do you think and if it's doable in the codebase without major changes.

Thanks.

**Comment by jagrosh** [comment]
This seems out-of-scope for this project. A web interface for an in-Discord music player sounds like a separate project entirely, and you'd probably find it easier to write it from scratch than try to create some sort of adapter for JMusicBot.

Regardless of your approach, I highly doubt anything like this would be added to JMusicBot, as it would undoubtedly add additional libraries (more file size), require refactors of most of the code, and add unnecessary complication/confusion for most users, and at worst-case, a security concern that users could potentially get tricked into configuring maliciously.

**Reply by weeeeelp** [reply]
I was looking to write the web UI part myself entirely and I fully agree that part is completely out of scope for JMusicBot itself.

The only thing I'm looking for is a way to manually invoke methods responsible for bot actions, via any way. However, if the bot is strongly built around the text channel message interaction way, I can imagine it would be quite a lot of effort to add an additional abstraction layer  to allow what I described. Thanks for the response.

**Reply by jagrosh** [reply]
I'm aware that you were only requesting the ability to run commands locally; all of my points were in relation to only that portion:


> add additional libraries (more file size)  

JMusicBot would need a new dependency to handle a local HTTP server or command-line server to listen for incoming commands, which would add additional size to the compiled jar.

> require refactors of most of the code

All of the commands, and some of the player handler code, would need to be refactored, as most of these rely on more from Discord than just the command message itself. All commands would also need to be abstracted and restructured to still fit within the confines of the existing command framework.

> add unnecessary complication/confusion for most users

Even if disabled by default, it's confusing to make the config file even larger than it already is, let alone something like this that would undoubtedly need several layers of configuration to be customizable.

> at worst-case, a security concern that users could potentially get tricked into configuring maliciously

Similar to the eval command (which already has very large warnings in the config file), messing up the configuration of such a feature could potentially allow any user of the internet full control of the bot.

**Reply by weeeeelp** [reply]
I had a peek trough the Commmand/MusicCommand/PlayCmd classes, and to my layman's understanding of Java it does look like a massive effort to rewrite to work outside of a Discord-event type context. That look trough the codebase is what I should have done initially before posting this. Also, I agree with your points about the security implications (re: local-only port bind) and the required configuration complexity.

Please disregard this idea and mark as solved - having put some more thought into it and reading your points I now realize just how big of a change it would be. I suppose software should do one thing well, and it would be unreasonable to request this from a solo maintainer. Thank you for the thorough explanation and for the bot project itself!
